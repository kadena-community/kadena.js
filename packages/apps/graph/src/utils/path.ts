import {
  mkdirSync,
  readdirSync,
  renameSync,
  rmdirSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from 'fs';
import { basename, join } from 'path';

export async function createDirAndWriteFile(
  dir: string,
  fileName: string,
  data: string,
): Promise<void> {
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, fileName), data);
}

export async function clearDir(dir: string, extension?: string): Promise<void> {
  const files = readdirSync(dir);
  for (const file of files) {
    const filePath = join(dir, file);
    try {
      if (statSync(filePath).isDirectory()) {
        await clearDir(filePath, extension); // Recursive call
        rmdirSync(filePath); // Remove the directory itself
      } else if (!extension || file.endsWith(extension)) {
        unlinkSync(filePath);
      }
    } catch (error) {
      console.error(`Error processing file ${filePath}: ${error}`);
    }
  }
}
export async function flattenFolder(
  basePath: string,
  fileExtensions: string[],
  currentPath: string = basePath,
): Promise<void> {
  // Read directory
  const files = readdirSync(currentPath);
  for (const file of files) {
    const filePath = join(currentPath, file);

    // if directory, recurse and delete folder
    if (statSync(filePath).isDirectory()) {
      await flattenFolder(basePath, fileExtensions, filePath);
      rmdirSync(filePath);
    } else if (fileExtensions.some((ext) => file.endsWith(ext))) {
      // if file, rename to include parent folder name
      if (basePath === currentPath) continue;
      const relativePath = currentPath.replace(basePath, '');
      const folderChain = relativePath.split('/').filter(Boolean).join('.');
      const newFilePath = join(
        basePath,
        `${basename(folderChain)}.${basename(file)}`,
      );
      renameSync(filePath, newFilePath);
    }
  }
}
