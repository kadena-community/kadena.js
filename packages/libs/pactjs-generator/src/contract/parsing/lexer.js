'use strict';
exports.__esModule = true;
exports.getLexerOutput =
  exports.getModuleAndMethods =
  exports.filterFlattenWrap =
  exports.wrap =
  exports.flattenRecursive =
  exports.lexer =
    void 0;
var moo = require('moo');
/**
 * @internal
 */
exports.lexer = moo.compile({
  decimal: /[-+]?(?:[0-9]+[.])?[0-9]+/,
  boolean: /true|false/,
  // https://pact-language.readthedocs.io/en/stable/pact-reference.html#atoms
  atom: {
    match: /[%#+\-_&$@<>=?*!|\/a-zA-Z]+[%#+\-_&$@<>=?*!|\/a-zA-Z0-9]*/,
    type: moo.keywords({
      bless: 'bless',
      defun: 'defun',
      defcap: 'defcap',
      defconst: 'defconst',
      defpact: 'defpact',
      defschema: 'defschema',
      deftable: 'deftable',
      let: 'let',
      let_star: 'let*',
      cond: 'cond',
      step: 'step',
      stepWithRollback: 'step-with-rollback',
      use: 'use',
      interface: 'interface',
      module: 'module',
      implements: 'implements',
    }),
  },
  symbol: /'[%#+\-_&$@<>=?*!|\/a-zA-Z0-9]+/,
  comment: { match: /;;?.*$/, lineBreaks: true },
  // ======== LITERALS ========
  dot: '.',
  lparen: '(',
  rparen: ')',
  lsquare: '[',
  rsquare: ']',
  comma: ',',
  lcurly: '{',
  rcurly: '}',
  dereference: '::',
  assign: ':=',
  colon: ':',
  // https://regex101.com/r/OVdomu/1/
  string: { match: /"(?:[^"\\]|\\.|\\\n)*"/, lineBreaks: true },
  ws: /[ \t]+/,
  nl: { match: /\n/, lineBreaks: true },
});
/**
 * @internal
 */
function flattenRecursive(arr) {
  return arr.reduce(function (acc, cur) {
    if (Array.isArray(cur)) {
      return acc.concat(flattenRecursive(cur));
    }
    return acc.concat(cur);
  }, []);
}
exports.flattenRecursive = flattenRecursive;
/**
 * @internal
 */
var wrap = function (type) {
  return function (a) {
    return {
      type: type,
      value: a,
    };
  };
};
exports.wrap = wrap;
/**
 * @internal
 */
var filterFlattenWrap = function (type) {
  return function (a) {
    return (0, exports.wrap)(type)(flattenRecursive(a).filter(Boolean));
  };
};
exports.filterFlattenWrap = filterFlattenWrap;
/**
 * @internal
 */
function getModuleAndMethods(contract, logger) {
  exports.lexer.reset(contract);
  var token = undefined;
  var state = {
    stack: [],
    moduleName: '',
    defunName: '',
    argName: '',
  };
  var KW_MODULE = 'module';
  var KW_DEFUN = 'defun';
  var KW_ATOM = 'atom';
  var STACK_DEFUN_TYPE = 'defun_type';
  var STACK_ARGS_TYPE = 'args_type';
  var STACK_ARGS = 'args';
  var lastThreeTokens = [undefined, undefined, undefined];
  var output = {};
  while ((token = exports.lexer.next())) {
    lastThreeTokens.shift();
    lastThreeTokens.push(token);
    logger(token);
    switch (token.type) {
      case KW_MODULE:
        state.stack.push(KW_MODULE);
        logger('push '.concat(KW_MODULE));
        break;
      case KW_ATOM:
        if (state.stack.includes(STACK_DEFUN_TYPE)) {
          output[state.moduleName].defuns[state.defunName].returnType =
            token.value;
          logger('pop '.concat(STACK_DEFUN_TYPE));
          state.stack.pop();
          break;
        }
        if (state.stack.includes(STACK_ARGS_TYPE)) {
          output[state.moduleName].defuns[state.defunName].args[
            state.argName
          ].type = token.value;
          logger('pop '.concat(STACK_ARGS_TYPE));
          state.stack.pop();
          break;
        }
        if (
          state.stack.includes(KW_MODULE) &&
          !state.stack.includes(KW_DEFUN) &&
          !state.stack.includes(STACK_ARGS)
        ) {
          // module
          if (!state.moduleName) {
            state.moduleName = token.value;
            output[token.value] = {
              name: token.value,
              defuns: {},
              line: token.line,
              col: token.col,
            };
          }
        }
        if (
          state.stack.includes(KW_MODULE) &&
          state.stack.includes(KW_DEFUN) &&
          state.defunName === '' &&
          !state.stack.includes(STACK_ARGS)
        ) {
          // module, defun
          if (output[state.moduleName].defuns[token.value] === undefined) {
            state.defunName = token.value;
            var method = token.value;
            var returnType = 'undefined';
            output[state.moduleName].defuns[method] = {
              method: method,
              returnType: returnType,
              line: token.line,
              col: token.col,
              args: {},
            };
          }
        } else if (
          state.stack.includes(KW_MODULE) &&
          state.stack.includes(KW_DEFUN) &&
          state.stack.includes(STACK_ARGS)
        ) {
          var line = token.line,
            col = token.col;
          output[state.moduleName].defuns[state.defunName].args[token.value] = {
            line: line,
            col: col,
            name: token.value,
            type: 'undefined',
          };
        } else if (
          state.stack.includes(KW_MODULE) &&
          state.stack.includes(KW_DEFUN) &&
          state.defunName === '' &&
          !state.stack.includes(STACK_ARGS)
        ) {
          if (output[state.moduleName].defuns[token.value] === undefined) {
            state.defunName = token.value;
            var method = token.value;
            output[state.moduleName].defuns[method] = {
              method: method,
              returnType: 'undefined',
              line: token.line,
              col: token.col,
              args: {},
            };
          }
        } else if (
          state.stack.includes(KW_MODULE) &&
          state.stack.includes(KW_DEFUN) &&
          state.stack.includes(STACK_ARGS)
        ) {
          output[state.moduleName].defuns[state.defunName].args[token.value] = {
            name: token.value,
            type: 'undefined',
            line: token.line,
            col: token.col,
          };
        }
        break;
      case 'colon':
        if (
          state.stack.includes(KW_DEFUN) &&
          !state.stack.includes(STACK_ARGS) &&
          lastThreeTokens[1].type === 'atom'
        ) {
          // in defun, but not in args
          // last one is atom
          logger('push '.concat(STACK_DEFUN_TYPE));
          state.stack.push(STACK_DEFUN_TYPE);
          break;
        }
        if (
          state.stack.includes(STACK_ARGS) &&
          lastThreeTokens[1].type === 'atom'
        ) {
          // in args
          // last one is atom
          state.stack.push(STACK_ARGS_TYPE);
          state.argName = lastThreeTokens[1].value;
          logger('push '.concat(STACK_ARGS_TYPE));
          break;
        }
        break;
      case KW_DEFUN:
        if (state.stack.includes(KW_MODULE)) {
          state.stack.push(KW_DEFUN);
          logger('push '.concat(KW_DEFUN));
        }
        break;
      case 'lparen':
        if (
          state.stack.includes(KW_DEFUN) &&
          !state.stack.includes(STACK_ARGS)
        ) {
          state.stack.push(STACK_ARGS);
          logger('push '.concat(STACK_ARGS));
        }
        break;
      case 'rparen':
        if (state.stack.includes(STACK_ARGS)) {
          // pop STACK_ARGS
          logger('pop '.concat(STACK_ARGS));
          logger('popping', state.stack[state.stack.length - 1]);
          state.stack.pop();
          // pop KW_DEFUN
          logger('pop '.concat(KW_DEFUN));
          logger('popping', state.stack[state.stack.length - 1]);
          state.stack.pop();
          // remove current defunName
          state.defunName = '';
        }
        break;
      default:
        break;
    }
  }
  return output;
}
exports.getModuleAndMethods = getModuleAndMethods;
var TYPED_TOKEN_MATCH = 'atom,colon,atom';
function lastThreeTokenTypedAtomMatch(tokens) {
  return (
    tokens
      .map(function (t) {
        return t === null || t === void 0 ? void 0 : t.type;
      })
      .join() === TYPED_TOKEN_MATCH
  );
}
var getLexerOutput = function (contract, logger) {
  if (logger === void 0) {
    logger = function () {};
  }
  exports.lexer.reset(contract);
  var token = undefined;
  var returnValue = [];
  while ((token = exports.lexer.next())) {
    logger(token);
    returnValue.push(token);
  }
  return returnValue;
};
exports.getLexerOutput = getLexerOutput;
