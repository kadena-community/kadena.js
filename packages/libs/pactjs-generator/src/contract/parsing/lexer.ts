import type { Lexer } from 'moo';
import * as moo from 'moo';

/**
 * @internal
 */
export const lexer: Lexer = moo.compile({
  decimal: /[-+]?(?:[0-9]+[.])?[0-9]+/,
  boolean: /true|false/,
  // https://pact-language.readthedocs.io/en/stable/pact-reference.html#atoms
  atom: {
    match: /[%#+\-_&$@<>=?*!|\/a-zA-Z]+[%#+\-_&$@<>=?*!|\/a-zA-Z0-9]*/,
    type: moo.keywords({
      bless: 'bless',
      defun: 'defun',
      defcap: 'defcap',
      defconst: 'defconst',
      defpact: 'defpact',
      defschema: 'defschema',
      deftable: 'deftable',
      let: 'let',
      let_star: 'let*',
      namespace: 'namespace',
      cond: 'cond',
      step: 'step',
      stepWithRollback: 'step-with-rollback',
      use: 'use',
      interface: 'interface',
      module: 'module',
      implements: 'implements',
    }),
  },
  symbol: /'[%#+\-_&$@<>=?*!|\/a-zA-Z0-9]+/,
  comment: { match: /;;?.*$/, lineBreaks: true },
  // ======== LITERALS ========
  dot: '.',
  lparen: '(',
  rparen: ')',
  lsquare: '[',
  rsquare: ']',
  comma: ',',
  lcurly: '{',
  rcurly: '}',
  dereference: '::',
  assign: ':=',
  colon: ':',
  // https://regex101.com/r/OVdomu/1/
  string: { match: /"(?:[^"\\]|\\.|\\\r?\n)*"/, lineBreaks: true },
  ws: /[ \t]+/,
  nl: { match: /\r?\n/, lineBreaks: true },
});

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
// export function flattenRecursive<T extends readonly any[]>(arr: T): unknown[] {
//   return arr.reduce((acc, cur) => {
//     if (Array.isArray(cur)) {
//       return acc.concat(flattenRecursive(cur));
//     }
//     return acc.concat(cur);
//   }, []);
// }

/**
 * @internal
 */
// export const wrap =
//   <T>(type: string): ((arr: T[]) => { type: string; value: T[] }) =>
//   (arr: T[]) => {
//     return {
//       type,
//       value: arr,
//     };
//   };

/**
 * @internal
 */
// export const filterFlattenWrap =
//   <T>(type: string): ((arr: T[]) => { type: string; value: unknown[] }) =>
//   (arr: T[]) =>
//     wrap(type)(flattenRecursive(arr).filter(Boolean));

/**
 * @internal
 */
export interface ILogger {
  (msg: object | string, ...args: Array<object | string>): void;
}

export const getLexerOutput = (
  contract: string,
  logger: ILogger = () => {},
): moo.Token[] => {
  lexer.reset(contract);
  let token: moo.Token | undefined = undefined;

  const returnValue: moo.Token[] = [];

  while ((token = lexer.next())) {
    logger(token);
    returnValue.push(token);
  }
  return returnValue;
};
