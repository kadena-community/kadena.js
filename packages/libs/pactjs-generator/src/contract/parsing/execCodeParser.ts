import { unwrapData } from './utils/dataWrapper';
import { functionCallParser } from './utils/functionCallParser';
import { getPointer } from './utils/getPointer';
import { FAILED } from './utils/parser-utilities';

/**
 * @alpha
 */
export interface IParsedCode {
  function: {
    module?: string;
    namespace?: string;
    name: string;
  };
  args: Array<
    | { string: string }
    | { int: number }
    | { decimal: number }
    | {
        object: Array<{ property: string; value: IParsedCode['args'][number] }>;
      }
    | { list: IParsedCode['args'] }
    | { code: IParsedCode }
  >;
}
/**
 * Parse the regular transaction code;
 *
 * this does not include deploy contract code; for that use {@link pactParser }
 * @example
 * const code = '(coin.transfer "alice" "bob" 100)(free.my-contract.my-function "alice" "bob" [100.1 2] \{ "extra" : "some-data" \} )';
 * const parsed = execCodeParser(code);
 * // const parsed = [
 * //   \{
 * //     function: \{ module: 'coin', name: 'transfer' \},
 * //     args: [\{ string: 'alice' \}, \{ string: 'bob' \}, \{ int: 100 \}],
 * //   \},
 * //   \{
 * //     function: \{
 * //       namespace: 'free',
 * //       module: 'my-contract',
 * //       name: 'my-function',
 * //     \},
 * //     args: [
 * //       \{ string: 'alice' \},
 * //       \{ string: 'bob' \},
 * //       \{ list: [\{ decimal: 100.1 \}, \{ int: 2 \}] \},
 * //       \{ object: [\{ property: 'extra', value: [\{ string: 'some-data' \}] \}] \},
 * //     ],
 * //   \},
 * // ];
 * @alpha
 */
export function execCodeParser(code: string): undefined | IParsedCode[] {
  const pointer = getPointer(code);
  const result = functionCallParser(pointer);
  const data = unwrapData(result);
  if (data === FAILED) {
    return undefined;
  }
  return data?.codes as undefined | IParsedCode[];
}
