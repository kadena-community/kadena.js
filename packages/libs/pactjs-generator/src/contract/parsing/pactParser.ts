import { unwrapData } from './utils/dataWrapper';
import { getCapabilities } from './utils/getCapabilities';
import type { IPointer } from './utils/getPointer';
import { getBlockPointer, getPointer } from './utils/getPointer';
import { functionCalls, parser } from './utils/pactGrammar';
import type { IModuleLike } from './utils/utils';
import { getModuleFullName } from './utils/utils';

export interface IType {
  kind: string;
  value: string;
  isList?: true;
}

interface ISchema {
  name: string;
  doc?: string;
  properties?: Array<{
    name: string;
    type: string | IType;
  }>;
}

interface IMethod {
  name: string;
  kind: string; // 'defun' | 'defcap' | 'defpact'; // need to fix typing
  returnType?: string | IType;
  doc?: string;
  parameters?: Array<{
    name: string;
    type: string | IType;
  }>;
}

export interface IFunction extends IMethod {
  bodyPointer?: number;
  requiredCapabilities?: string[];
  withCapabilities?: string[];
  functionCalls?: {
    internal: string[];
    external: Array<{ namespace?: string; module: string; func: string }>;
  };
  externalFnCalls?: Array<{
    namespace?: string;
    module: string;
    func: string;
  }>;
  allExtractedCaps?: Array<{
    name: string;
    fullModuleName: string;
    reason: 'with-capability' | 'compose-capability';
    origin: string;
    capability: ICapability;
  }>;
}

export interface ICapability extends IMethod {
  composeCapabilities?: string[];
}

export interface IModule {
  kind: string;
  namespace: string;
  name: string;
  doc: string;
  governance: string;
  usedModules?: Array<{
    name: string;
    namespace?: string;
    hash?: string;
    imports?: string[];
  }>;
  usedInterface?: Array<{ name: string; namespace?: string }>;
  functions?: IFunction[];
  capabilities?: ICapability[];
  schemas?: ISchema[];
}

interface IUsedModules {
  name: string;
  hash?: string;
  location?: number;
  namespace?: string;
  imports?: string[];
}

interface IPactTree {
  namespace?: string[];
  usedModules?: Array<IUsedModules>;
  module?: IModule[];
}

// return the namespace of the module in case the file has multiple modules and namespaces
const getNamespace = (
  moduleLoc: number,
  allNamespaces?: Array<{ location: number; name: string }>,
): string => {
  if (allNamespaces === undefined) return '';
  const { name } = allNamespaces.reduce(
    (prev, namespace) => {
      if (
        namespace.location < moduleLoc &&
        namespace.location > prev.location
      ) {
        return namespace;
      }
      return prev;
    },
    { location: -2, name: '' },
  );
  return name;
};

// return the list of modules used in the module including the modules used in root level
const getUsedModules = (
  moduleLoc: number,
  allUsedModules?: IUsedModules[],
): Array<Omit<IUsedModules, 'location'>> => {
  if (!allUsedModules) return [];
  const list = allUsedModules.reduce(
    (list, { location, ...mod }) => {
      if (location !== undefined && location < moduleLoc) {
        return [...list, mod];
      }
      return list;
    },
    [] as Exclude<IPactTree['usedModules'], undefined>,
  );

  return list;
};

// returns the list of modules used in the functions of the module without using "use" keyword in the module
function getUsedModulesInFunctions(
  functions?: IFunction[],
): Required<IModuleLike>[] {
  if (!functions) return [];

  return functions.flatMap((fun) => {
    const externalFnCalls = fun.externalFnCalls;
    if (externalFnCalls === undefined) return [];
    return externalFnCalls.map(({ namespace, module: name }) => ({
      namespace: namespace ?? '',
      name,
    }));
  });
}
type IsNotDuplicated = <T>(
  isEqual: (a: T, b: T) => boolean,
) => (a: T, idx: number, list: T[]) => boolean;

const isNotDuplicated: IsNotDuplicated =
  <T>(isEqual: (a: T, b: T) => boolean) =>
  (a: T, idx: number, list: T[]): boolean =>
    list.findIndex((b) => isEqual(a, b)) === idx;

// eslint-disable-next-line @rushstack/typedef-var
const isNotDuplicatedModule = isNotDuplicated<IModuleLike>(
  (a, b) => a.name === b.name && a.namespace === b.namespace,
);

/**
 * @alpha
 */
export function contractParser(
  contract: string,
  namespace: string = '',
): [IModule[], IPointer] {
  const pointer = getPointer(contract);
  const result = parser(pointer);
  const tree = unwrapData(result);
  const { modules, namespaces, usedModules } = tree;

  const extModules =
    modules !== undefined
      ? modules.map(({ location, module: mod }) => ({
          ...mod,
          namespace: getNamespace(location, namespaces) || namespace,
          usedModules: [
            ...(mod.usedModules ?? []),
            ...getUsedModules(location, usedModules),
            ...getUsedModulesInFunctions(mod.functions),
          ].filter(isNotDuplicatedModule),
        }))
      : [];

  return [extModules, pointer];
}

interface IModuleLoader {
  getModule(moduleFullName: string): Promise<IModuleWithPointer | undefined>;
  parserFile(content: string, namespace?: string): void;
  getStorage(): Map<string, IModuleWithPointer>;
}

const moduleLoader = (
  fetchModule: (moduleFullName: string) => Promise<string>,
): IModuleLoader => {
  const storage = new Map<string, IModuleWithPointer>();

  const parseModule = (content: string, namespace?: string): void => {
    const [modules, pointer] = contractParser(content, namespace);
    modules.forEach((mod) => {
      storage.set(getModuleFullName(mod), { ...mod, pointer });
    });
  };

  const cache = new Map<string, Promise<string>>();
  const fetchContract = (name: string): Promise<string> => {
    if (cache.has(name)) return cache.get(name)!;
    const pr = fetchModule(name);
    cache.set(name, pr);
    return pr;
  };

  return {
    async getModule(moduleFullName: string) {
      const slugs = moduleFullName.split('.');
      const namespace = slugs.length === 2 ? slugs[0] : '';

      const mod = storage.get(moduleFullName);
      if (mod !== undefined) return mod;

      const content = await fetchContract(moduleFullName);
      if (content) {
        parseModule(content, namespace);
      }

      return storage.get(moduleFullName)!;
    },
    parserFile(content: string, namespace: string) {
      parseModule(content, namespace);
    },
    getStorage() {
      return storage;
    },
  };
};

interface IModuleWithPointer extends IModule {
  pointer?: IPointer;
}

// parse a module and its dependencies
async function loadModuleDependencies(
  fullName: string,
  getModule: (name: string) => Promise<IModuleWithPointer | undefined>,
): Promise<Array<IModuleWithPointer>> {
  const module = await getModule(fullName);
  // skip this module if its not available
  if (!module) return [];

  const interfaceOrModule = [
    ...(module.usedModules || []),
    ...(module.usedInterface || []),
  ];

  const mods = await Promise.all(
    interfaceOrModule.map(async (usedModule) =>
      loadModuleDependencies(getModuleFullName(usedModule), getModule),
    ),
  );
  const dependencies = mods.flat();

  return [module, ...dependencies].filter(isNotDuplicatedModule);
}

// check function body again for checking function calls, internal and external
const getFunctionCalls = (
  bodyPointer: number | undefined,
  pointer: IPointer,
  fnList: string[],
  availableExternals: Array<{
    namespace?: string;
    module: string;
    func: string;
  }>,
): {
  internal?: string[];
  external?: Array<{ namespace?: string; module: string; func: string }>;
} => {
  if (bodyPointer === undefined) return {};

  pointer.reset(bodyPointer);
  const blockPointer = getBlockPointer(pointer);
  const data = functionCalls(fnList, availableExternals)(blockPointer);

  return unwrapData(data);
};

// adding function calls to each function in the module
function addFunctionCalls(
  mod: IModule,
  allModule: Map<string, IModule>,
  pointer: IPointer,
): void {
  if (mod.kind !== 'module') return;

  const usedModules = mod.usedModules;
  let externalFunctions: Array<{
    module: string;
    namespace: string | undefined;
    func: string;
  }> = [];

  if (usedModules && usedModules.length) {
    externalFunctions = usedModules.flatMap(({ name, namespace, imports }) => {
      if (imports && imports.length) {
        return imports.map((func) => ({
          module: name,
          namespace,
          func,
        }));
      }
      const usedModule = allModule.get(getModuleFullName({ name, namespace }));
      if (!usedModule || !usedModule.functions) return [];
      return usedModule.functions.map(({ name: func }) => ({
        module: name,
        namespace,
        func,
      }));
    });
  }
  if (!mod.functions) {
    mod.functions = [];
  }
  const internalFunctions = mod.functions.map(({ name }) => name);
  mod.functions = mod.functions.map((fun) => {
    const { internal = [], external = [] } = getFunctionCalls(
      fun.bodyPointer,
      pointer,
      internalFunctions,
      externalFunctions,
    );

    fun.functionCalls = {
      internal,
      external: [...external, ...(fun.externalFnCalls || [])],
    };

    delete fun.bodyPointer;

    return fun;
  });
}

// extract all capabilities for each function in the module
function addFunctionCapabilities(
  mod: IModule,
  allModule: Map<string, IModule>,
): void {
  if (mod.kind !== 'module' || !mod.functions) return;
  mod.functions.forEach((fun) => {
    fun.allExtractedCaps = getCapabilities(
      allModule,
      getModuleFullName(mod),
      fun.name,
    );
  });
}

/**
 * @alpha
 */
export async function pactParser({
  contractNames,
  files,
  getContract,
  namespace,
}: {
  contractNames?: string[];
  files?: string[];
  getContract: (fullName: string) => Promise<string>;
  namespace?: string;
}): Promise<{ [k: string]: IModule }> {
  const loader = moduleLoader(getContract);

  // parse files if presented
  if (files !== undefined) {
    files.forEach((content) => {
      loader.parserFile(content, namespace);
    });
  }

  // fetch and parse contracts by name if presented
  if (contractNames !== undefined) {
    await Promise.all(
      contractNames.map((contract) => {
        return loader.getModule(contract);
      }),
    );
  }

  const parsedModules = loader.getStorage();

  if (parsedModules.size === 0) {
    throw new Error('NO_MODULE_LOADED');
  }

  // load all dependencies
  await Promise.all(
    [...parsedModules.keys()].map((name) => {
      return loadModuleDependencies(name, loader.getModule);
    }),
  );

  const allModules = loader.getStorage();

  allModules.forEach((mod) => {
    addFunctionCalls(mod, allModules, mod.pointer!);
    delete mod.pointer;
  });

  allModules.forEach((mod) => {
    addFunctionCapabilities(mod, allModules);
  });

  return Object.fromEntries(allModules);
}
