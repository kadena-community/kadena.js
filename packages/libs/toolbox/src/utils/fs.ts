import { mkdir, stat, writeFile as writeFileN } from 'node:fs/promises';
import { dirname } from 'node:path';

export const writeFileAtPath: typeof writeFileN = async (
  path,
  data,
  options,
) => {
  if (typeof path !== 'string') throw new Error('path must be a string');
  const exist = await stat(path)
    .then(() => true)
    .catch(() => false);
  if (!exist) {
    const dir = dirname(path);
    await mkdir(dir, { recursive: true });
  }
  return writeFileN(path, data, options);
};
