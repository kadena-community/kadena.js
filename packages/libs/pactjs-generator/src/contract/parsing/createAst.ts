import { type ILogger, lexer } from './lexer';

export type IAst = IAstNode[];
export type IAstNode = moo.Token & {
  children?: IAstNode[];
};

const LPAREN: 'lparen' = 'lparen';
const RPAREN: 'rparen' = 'rparen';

export function createAst(contract: string, logger: ILogger = () => {}): IAst {
  lexer.reset(contract);
  return getChildren(lexer, logger);
}

function getChildren(
  lexer: moo.Lexer,
  logger: ILogger,
  parentNodes: IAstNode[] = [],
): IAstNode[] {
  let token: moo.Token | undefined;
  const nodes: IAstNode[] = [];
  while ((token = lexer.next())) {
    switch (token.type) {
      case 'nl':
      case 'ws':
        break;

      case LPAREN:
        const newNode: IAstNode = { ...token };
        nodes.push(newNode);
        newNode.children = getChildren(lexer, logger, nodes);
        break;

      case RPAREN:
        parentNodes.push(token);
        return nodes;

      default:
        nodes.push(token);
        break;
    }
  }

  return nodes;
}

// [ token1, token2 ]
