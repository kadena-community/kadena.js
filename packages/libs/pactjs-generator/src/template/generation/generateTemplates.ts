import { ITemplate } from '../parsing/parseTemplate';

function uniq(acc: string[], curr: string): string[] {
  if (acc.includes(curr)) {
    return acc;
  }
  acc.push(curr);
  return acc;
}

const commaBetweenQuotesRegex: RegExp = /","/g;
const closingBracketAtEndOfStringRegex: RegExp = /]$/;

function fixArrayFormatting(arrAsString: string): string {
  return arrAsString
    .replace('[', `[\n      `)
    .replace(commaBetweenQuotesRegex, `",\n      "`)
    .replace(closingBracketAtEndOfStringRegex, '\n    ]');
}

/**
 * @alpha
 */
export function generateTemplates(
  templates: { name: string; template: ITemplate }[],
  version: string,
): string {
  return (
    `// generated by pactjs-generator and pactjs-cli@${version}
import { buildUnsignedTransaction, ICommandBuilder } from '@kadena/client';
` +
    // Rationale: export default because we cannot export "some-template", and
    // we don't want to modify the accessors (e.g. convert to camelCase)
    `
export default {${templates
      .map(
        (t) => `
  "${t.name}": (args: {${t.template.holes
    .reduce(uniq, [] as string[])
    .map(
      (h) => `
    "${h}": string,`,
    )
    .join('')}
  }): ICommandBuilder<{}> => {
    const parts = ${fixArrayFormatting(JSON.stringify(t.template.parts))};
    const holes = ${fixArrayFormatting(JSON.stringify(t.template.holes))};

    return buildUnsignedTransaction(parts, holes, args);
  }`,
      )
      .join(',\n  ')}
}`
  );
}
