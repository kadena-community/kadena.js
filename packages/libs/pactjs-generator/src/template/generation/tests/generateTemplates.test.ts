import { readFileSync } from 'fs';
import { join } from 'path';
import { parseTemplate } from '../../parsing/parseTemplate';
import { generateTemplates } from '../generateTemplates';

describe('generateDts', () => {
  it('generates a ts for `Hello, ${name}!`', () => {
    const ts = generateTemplates(
      [
        {
          name: 'hello-name',
          template: { parts: ['Hello, ', '!'], holes: ['name'] },
        },
      ],
      '0.0.1-alpha-test',
    );

    expect(ts).toMatchSnapshot();
  });

  it('generates a ts for a full template', () => {
    const simpleTransferTpl = parseTemplate(
      readFileSync(
        join(__dirname, '../my-tx-lib/src/simple-transfer.json'),
        'utf8',
      ),
    );

    const safeTransactionTpl = parseTemplate(
      readFileSync(
        join(__dirname, '../my-tx-lib/src/safe-transaction.yaml'),
        'utf8',
      ),
    );

    const ts = generateTemplates(
      [
        {
          name: 'simple-transfer',
          template: simpleTransferTpl,
        },
        {
          name: 'safe-transaction',
          template: safeTransactionTpl,
        },
      ],
      '0.0.1-alpha-test',
    );

    expect(ts).toMatchSnapshot();
  });
});
