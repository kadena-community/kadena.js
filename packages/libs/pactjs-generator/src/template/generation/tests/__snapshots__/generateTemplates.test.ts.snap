// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDts generates a ts for \`Hello, \${name}!\` 1`] = `
"// generated by pactjs-generator and pactjs-cli@0.0.1-alpha-test
import { buildUnsignedTransaction, ICommandBuilder } from '@kadena/client';

export default {
  "hello-name": (args: {
    "name": string,
  }): ICommandBuilder<{}> => {
    const parts = [
      "Hello, ",
      "!"
    ];
    const holes = [
      "name"
    ];

    return buildUnsignedTransaction(parts, holes, args);
  }
}"
`;

exports[`generateDts generates a ts for a full template 1`] = `
"// generated by pactjs-generator and pactjs-cli@0.0.1-alpha-test
import { buildUnsignedTransaction, ICommandBuilder } from '@kadena/client';

export default {
  "simple-transfer": (args: {
    "from-acct": string,
    "to-acct": string,
    "amount": string,
    "chain": string,
    "network": string,
    "from-key": string,
  }): ICommandBuilder<{}> => {
    const parts = [
      "{\\n  \\"code\\": \\"(coin.transfer \\\\\\"",
      "\\\\\\" \\\\\\"",
      "\\\\\\" ",
      ")\\",\\n  \\"data\\": null,\\n  \\"publicMeta\\": {\\n    \\"chainId\\": \\"",
      "\\",\\n    \\"sender\\": \\"",
      "\\",\\n    \\"gasLimit\\": 2500,\\n    \\"gasPrice\\": 1e-8,\\n    \\"ttl\\": 600\\n  },\\n  \\"networkId\\": \\"",
      "\\",\\n  \\"signers\\": [\\n    {\\n      \\"public\\": \\"",
      "\\",\\n      \\"caps\\": [\\n        { \\"name\\": \\"coin.GAS\\", \\"args\\": [] },\\n        {\\n          \\"name\\": \\"coin.TRANSFER\\",\\n          \\"args\\": [\\"",
      "\\", \\"",
      "\\", \\"",
      "\\"]\\n        }\\n      ]\\n    }\\n  ],\\n  \\"type\\": \\"exec\\"\\n}\\n"
    ];
    const holes = [
      "from-acct",
      "to-acct",
      "amount",
      "chain",
      "from-acct",
      "network",
      "from-key",
      "from-acct",
      "to-acct",
      "amount"
    ];

    return buildUnsignedTransaction(parts, holes, args);
  },

  "safe-transaction": (args: {
    "{from-acct": string,
    "{to-acct": string,
    "amount": string,
    "chain": string,
    "network": string,
    "from-key": string,
  }): ICommandBuilder<{}> => {
    const parts = [
      "code: |-\\n  (coin.transfer \\"",
      "}\\" \\"",
      "}\\" ",
      ")\\ndata:\\npublicMeta:\\n  chainId: '",
      "'\\n  sender: '",
      "}'\\n  gasLimit: 600\\n  gasPrice: 0.00000001\\n  ttl: 7200\\nnetworkId: '",
      "'\\nsigners:\\n  - public: '",
      "'\\n    caps:\\n      - name: 'coin.TRANSFER'\\n        args:\\n          - '",
      "}'\\n          - '",
      "}'\\n          - '",
      "'\\n      - name: 'coin.GAS'\\n        args: []\\ntype: exec\\n"
    ];
    const holes = [
      "{from-acct",
      "{to-acct",
      "amount",
      "chain",
      "{from-acct",
      "network",
      "from-key",
      "{from-acct",
      "{to-acct",
      "amount"
    ];

    return buildUnsignedTransaction(parts, holes, args);
  }
}"
`;
